crates/apps/src/lib/node/ledger/shell/mod.rs

```diff
@@ -1486,67 +1473,95 @@ where
     // further validation
 
     // Validate data and generate unshielding tx
-    let transfer_code_hash = get_transfer_hash_from_storage(temp_wl_storage);
-
-    let descriptions_limit = temp_wl_storage
-        .read(
-            &parameters::storage::get_fee_unshielding_descriptions_limit_key(),
-        )
-        .expect("Error reading the storage")
-        .expect("Missing fee unshielding descriptions limit param in storage");
+    check_fee_unshielding(shell_params.wl_storage, wrapper, &masp_transaction)?;
 
-    let unshield = wrapper
-        .check_and_generate_fee_unshielding(
-            transfer_code_hash,
-            Some(namada_sdk::tx::TX_TRANSFER_WASM.to_string()),
-            descriptions_limit,
-            masp_transaction,
-        )
-        .map_err(|e| Error::TxApply(protocol::Error::FeeUnshieldingError(e)))?;
-
-    let fee_unshielding_gas_limit: GasLimit = temp_wl_storage
-        .read(&parameters::storage::get_fee_unshielding_gas_limit_key())
-        .expect("Error reading from storage")
-        .expect("Missing fee unshielding gas limit in storage");
-
-    // Runtime check
-    // NOTE: A clean tx write log must be provided to this call for a
-    // correct vp validation. Block write log, instead, should contain
-    // any prior changes (if any). This is to simulate the
-    // unshielding tx (to prevent the already written keys
-    // from being passed/triggering VPs) but we cannot
-    // commit the tx write log yet cause the tx could still
-    // be invalid.
-    temp_wl_storage.write_log.precommit_tx();
-
-    let result = apply_wasm_tx(
-        unshield,
-        &TxIndex::default(),
-        ShellParams::new(
-            &mut TxGasMeter::new(fee_unshielding_gas_limit),
-            temp_wl_storage,
-            vp_wasm_cache,
-            tx_wasm_cache,
-        ),
+    if namada::ledger::protocol::run_fee_unshielding(
```

Comment:
I'm now using for `mempool_validate`, `prepare_proposal` and `process_proposal` the same function used in `finalize_block`. The logic though is a bit different: in the first three cases, if the fee unshielding operation fails, for any reason, the transaction is deemed invalid (won't make it into the block). In `finalize_block`, instead, if the fee unshielding fails we still try to get the fee payment from the transparent address of the fee payer: if the funds are not enough we take what we can and reject the tx (won't get executed) otherwise, at least for now, the transaction can go ahead.

It has to be noted that this scenario in `finalize_block` should never happen, the check in `process_proposal` should prevent such a transaction from being included into the block. We still need to handle this case in case there was a bug and a transaction of this kind could indeed be included. We could also choose, for this case, to still try to get the payment from the transparent address but reject the transaction even if the payment was successful.